qbft/spectest/tests/controller/decided/late_decided.go

```diff
@@ -25,8 +26,55 @@ func LateDecided() tests.SpecTest {
 					DecidedVal:         testingutils.TestingQBFTFullData,
 					BroadcastedDecided: testingutils.TestingCommitMultiSignerMessage([]*bls.SecretKey{ks.Shares[1], ks.Shares[2], ks.Shares[3]}, []types.OperatorID{1, 2, 3}),
 				},
-				ControllerPostRoot: "4c96913e87aa17c9f0d5c1b6b220cbc7a66b7b40ef55f1059f1e1fa9f59c94d9",
+				ControllerPostRoot: lateDecidedStateComparison().Register().Root(),
 			},
 		},
 	}
 }
+
+func lateDecidedStateComparison() *qbftcomparable.StateComparison {
+	identifier := []byte{1, 2, 3, 4}
+	config := testingutils.TestingConfig(testingutils.Testing4SharesSet())
+	contr := testingutils.NewTestingQBFTController(
+		identifier[:],
+		testingutils.TestingShare(testingutils.Testing4SharesSet()),
+		config,
+	)
+	_ = contr.StartNewInstance(testingutils.TestingQBFTFullData)
+
+	ks := testingutils.Testing4SharesSet()
+
+	msgs := testingutils.DecidingMsgsForHeightWithRoot(testingutils.TestingQBFTRootData, testingutils.TestingQBFTFullData, testingutils.TestingIdentifier, qbft.FirstHeight, ks)
+	msgs = append(msgs, testingutils.TestingCommitMultiSignerMessage([]*bls.SecretKey{ks.Shares[1], ks.Shares[2], ks.Shares[4]}, []types.OperatorID{1, 2, 4}))
+
+	state := testingutils.BaseInstance().State
+	state.ProposalAcceptedForCurrentRound = testingutils.TestingProposalMessage(ks.Shares[1], types.OperatorID(1))
+	state.LastPreparedRound = 1
+	state.LastPreparedValue = testingutils.TestingQBFTFullData
+	state.Decided = true
+	state.DecidedValue = testingutils.TestingQBFTFullData
+
+	state.ProposeContainer = &qbft.MsgContainer{Msgs: map[qbft.Round][]*qbft.SignedMessage{
+		qbft.FirstRound: {
+			msgs[0],
```

Comment:
for readability maybe better to have the actual message written

Reply:
It will add more unnecessary work
we are using already build-in msgs to build the state
https://github.com/olegshmuelov/ssv-spec/blob/3a7d5264669f26613f5cade076b50d8c84b7a954/types/testingutils/runner.go#L231

Reply:
you have a circular dependency here, if you change something in function creating the array of messages the test will change as well..

Reply:
this specific test using [this function](https://github.com/olegshmuelov/ssv-spec/blob/3a7d5264669f26613f5cade076b50d8c84b7a954/qbft/spectest/tests/controller/decided/late_decided.go#L16) for input messages
if it will be changed for actual test, it should be changed for expected msgs as well.

agree that it can be changed for better readability.
but in some cases e.g. in [/late_prepare_past_instance.go](https://github.com/olegshmuelov/ssv-spec/blob/3a7d5264669f26613f5cade076b50d8c84b7a954/qbft/spectest/tests/controller/latemsg/late_prepare_past_instance.go#L24) we have about 42 messages.
@GalRogozinski

Reply:
maybe what we can do is add `expectedDecidedMessages` in testutils that have return the array of decided message.
This will keep the current design and break the circular dependency

Reply:
In general it might be a good idea to do the same even with simple constants (even though one can argue it is an "overkill")
The idea should be that the dev should always change the test and the expected state separately

Reply:
https://github.com/bloxapp/ssv-spec/pull/190/commits/2208a961c8eb0f7608e9ef1bfa0d653b322fc46f
https://github.com/bloxapp/ssv-spec/pull/190/commits/7658796119336c2b8e0c538c2ee2534e338fcc17

added expectedDecidedMessages
and made the code cleaner

Reply:
I still have an issue with using testingutils.DecidingMsgsForHeightWithRoot for the state comparison, we are creating circular dependency which cuts against the state comparison. If something happened inside DecidingMsgsForHeightWithRoot the test can be wrong but still pass

Reply:
@alonmuroch take a look on my last comment with the commits links
we are using expectedDecidedMessages as requested.
You are looking on outdated file again

Reply:
I don't thin I do... You are getting messages from testingutils.DecidingMsgsForHeightWithRoot and putting them blindly (you don't know what message is in index 1/2/3/etc... 
If testingutils.DecidingMsgsForHeightWithRoot produced the wrong messages you won't know about it

Reply:
Ok, I just fell for that github bug as well...
Apparently @olegshmuelov is correct that it is outdated. But sometimes Github doesn't show it.